File: ueueo-framework/ueueo-authorization/src/main/java/com/ueueo/authorization/permissions/PermissionDefinitionManager.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Volo.Abp.DependencyInjection;

namespace Volo.Abp.Authorization.Permissions;

public class PermissionDefinitionManager : IPermissionDefinitionManager, ISingletonDependency
{
    protected IDictionary<String, PermissionGroupDefinition> PermissionGroupDefinitions => _lazyPermissionGroupDefinitions.Value;
    private readonly Lazy<Dictionary<String, PermissionGroupDefinition>> _lazyPermissionGroupDefinitions;

    protected IDictionary<String, PermissionDefinition> PermissionDefinitions => _lazyPermissionDefinitions.Value;
    private readonly Lazy<Dictionary<String, PermissionDefinition>> _lazyPermissionDefinitions;

    protected AbpPermissionOptions Options;//  { get; }

    private readonly IServiceProvider _serviceProvider;

    public PermissionDefinitionManager(
        IOptions<AbpPermissionOptions> options,
        IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
        Options = options.Value;

        _lazyPermissionDefinitions = new Lazy<Dictionary<String, PermissionDefinition>>(
            CreatePermissionDefinitions,
            isThreadSafe: true
        );

        _lazyPermissionGroupDefinitions = new Lazy<Dictionary<String, PermissionGroupDefinition>>(
            CreatePermissionGroupDefinitions,
            isThreadSafe: true
        );
    }

    public   PermissionDefinition Get(String name)
    {
        var permission = GetOrNull(name);

        if (permission == null)
        {
            throw new AbpException("Undefined permission: " + name);
        }

        return permission;
    }

    public   PermissionDefinition GetOrNull(String name)
    {
        Objects.requireNonNull(name, nameof(name));

        return PermissionDefinitions.GetOrDefault(name);
    }

    public   IReadOnlyList<PermissionDefinition> GetPermissions()
    {
        return PermissionDefinitions.Values.ToImmutableList();
    }

    public IReadOnlyList<PermissionGroupDefinition> GetGroups()
    {
        return PermissionGroupDefinitions.Values.ToImmutableList();
    }

    protected   Dictionary<String, PermissionDefinition> CreatePermissionDefinitions()
    {
        var permissions = new Dictionary<String, PermissionDefinition>();

        for (var groupDefinition in PermissionGroupDefinitions.Values)
        {
            for (var permission in groupDefinition.Permissions)
            {
                AddPermissionToDictionaryRecursively(permissions, permission);
            }
        }

        return permissions;
    }

    protected   void AddPermissionToDictionaryRecursively(
        Dictionary<String, PermissionDefinition> permissions,
        PermissionDefinition permission)
    {
        if (permissions.ContainsKey(permission.Name))
        {
            throw new AbpException("Duplicate permission name: " + permission.Name);
        }

        permissions[permission.Name] = permission;

        for (var child in permission.Children)
        {
            AddPermissionToDictionaryRecursively(permissions, child);
        }
    }

    protected   Dictionary<String, PermissionGroupDefinition> CreatePermissionGroupDefinitions()
    {
        using (var scope = _serviceProvider.CreateScope())
        {
            var context = new PermissionDefinitionContext(scope.ServiceProvider);

            var providers = Options
                    .DefinitionProviders
                    .Select(p => scope.ServiceProvider.GetRequiredService(p) as IPermissionDefinitionProvider)
                    .ToList();

            for (var provider in providers)
            {
                provider.PreDefine(context);
            }

            for (var provider in providers)
            {
                provider.Define(context);
            }

            for (var provider in providers)
            {
                provider.PostDefine(context);
            }

            return context.Groups;
        }
    }
}

File: ueueo-framework/ueueo-authorization/src/main/java/com/ueueo/authorization/permissions/PermissionDefinitionManager.java
package com.ueueo.authorization.permissions;

import com.ueueo.exception.SystemException;
import org.springframework.beans.factory.BeanFactory;

import java.util.*;
import java.util.stream.Collectors;

public class PermissionDefinitionManager implements IPermissionDefinitionManager {
    protected Map<String, PermissionGroupDefinition> permissionGroupDefinitions;

    protected Map<String, PermissionDefinition> permissionDefinitions;

    protected AbpPermissionOptions options;

    private BeanFactory beanFactory;

    public PermissionDefinitionManager(
            AbpPermissionOptions options,
            BeanFactory beanFactory) {
        this.beanFactory = beanFactory;
        this.options = options;
        this.permissionDefinitions = new HashMap<>();
        this.permissionDefinitions.putAll(createPermissionDefinitions());
        this.permissionGroupDefinitions = new HashMap<>();
        this.permissionGroupDefinitions.putAll(createPermissionGroupDefinitions());
    }

    @Override
    public PermissionDefinition get(String name) {
        PermissionDefinition permission = getOrNull(name);

        if (permission == null) {
            throw new SystemException("Undefined permission: " + name);
        }

        return permission;
    }

    @Override
    public PermissionDefinition getOrNull(String name) {
        Objects.requireNonNull(name);

        return permissionDefinitions.get(name);
    }

    @Override
    public List<PermissionDefinition> getPermissions() {
        return new ArrayList<>(permissionDefinitions.values());
    }

    @Override
    public List<PermissionGroupDefinition> getGroups() {
        return new ArrayList<>(permissionGroupDefinitions.values());
    }

    protected Map<String, PermissionDefinition> createPermissionDefinitions() {
        Map<String, PermissionDefinition> permissions = new HashMap<>();

        for (PermissionGroupDefinition groupDefinition : permissionGroupDefinitions.values()) {
            for (PermissionDefinition permission : groupDefinition.getPermissions()) {
                addPermissionToDictionaryRecursively(permissions, permission);
            }
        }

        return permissions;
    }

    protected void addPermissionToDictionaryRecursively(
            Map<String, PermissionDefinition> permissions,
            PermissionDefinition permission) {
        if (permissions.containsKey(permission.getName())) {
            throw new SystemException("Duplicate permission name: " + permission.getName());
        }
        permissions.put(permission.getName(), permission);

        for (PermissionDefinition child : permission.getChildren()) {
            addPermissionToDictionaryRecursively(permissions, child);
        }
    }

    protected Map<String, PermissionGroupDefinition> createPermissionGroupDefinitions() {

        PermissionDefinitionContext context = new PermissionDefinitionContext();

        List<IPermissionDefinitionProvider> providers = options
                .getDefinitionProviders()
                .stream().map(beanFactory::getBean)
                .collect(Collectors.toList());

        for (IPermissionDefinitionProvider provider : providers) {
            provider.preDefine(context);
        }

        for (IPermissionDefinitionProvider provider : providers) {
            provider.define(context);
        }

        for (IPermissionDefinitionProvider provider : providers) {
            provider.postDefine(context);
        }

        return context.getGroups();

    }
}
